# 单元测试
[参考网站：廖](https://www.liaoxuefeng.com/wiki/1016959663602400/1017604210683936#0)


**单元测试是用来对一个模块、一个函数或者一个类来进行正确性检验的测试工作**

> ***1. 需要测试用例和测试模块***  
> ***2. 将测试用例放到一个测试模块***  
**3. 如果单元测试通过，说明我们测试的这个函数能够正常工作。如果单元测试不通过，要么函数有bug，要么测试条件输入不正确，总之，需要修复使单元测试能够通过**

## 一、单元测试时如何运作的
例如：

1. 编写一个Dict类，这个类的行为和dict一致，但是可以通过属性来访问，用起来就效果如下：
```python
>>> d = Dict(a=1, b=2)
>>> d['a']
1
>>> d.a
1
```


2. 测试用例mydict.py代码如下：
```python
class Dict(dict):
    '''此为测试用例mydict.py代码'''
    def __init__(self, **kw):
        super().__init__(**kw)

    def __getattr__(self, key):
        try:
            return self[key]
        except KeyError:
            raise AttributeError(r"'Dict' object has no attribute '%s'" % key)

    def __setattr__(self, key, value):
        self[key] = value
```

3.测试模块代码如下：
```python
import unittest   # 导包

from mydict import Dict

class TestDict(unittest.TestCase):   # 编写单元测试时，我们需要编写一个测试类，从unittest.TestCase继承
    '''此为测试模块mydict_test.py代码'''

    def test_init(self):   #以test开头的方法就是测试方法，不以test开头的方法不被认为是测试方法，测试的时候不会被执行
        d = Dict(a=1, b='test')
        self.assertEqual(d.a, 1)
        self.assertEqual(d.b, 'test')
        self.assertTrue(isinstance(d, dict))

    def test_key(self):   # 对每一类测试都需要编写一个test_xxx()方法
        d = Dict()
        d['key'] = 'value'
        self.assertEqual(d.key, 'value')

    def test_attr(self):
        d = Dict()
        d.key = 'value'
        self.assertTrue('key' in d)
        self.assertEqual(d['key'], 'value')

    def test_keyerror(self):
        d = Dict()
        with self.assertRaises(KeyError):
            value = d['empty']

    def test_attrerror(self):
        d = Dict()
        with self.assertRaises(AttributeError):
            value = d.empty
```


## 二、unittest.TestCase常用的两个断言：

1. assertEqual()
判断两者是否相等
例如：  
```python
self.assertEqual(abs(-1), 1) # 断言函数返回的结果与1相等
```

2. assertRaises()  

期待抛出指定类型的Error  
例如1：  
通过`d['empty']`访问不存在的`key`时，断言会抛出`KeyError`  
```python
with self.assertRaises(KeyError):
    value = d['empty']
```
例如2：  
通过`d.empty`访问不存在的key时，我们期待抛出`AttributeError`
```python
with self.assertRaises(AttributeError):
    value = d.empty
```



## 三、对单元测试的补充：setUp与tearDown


`setUp()`和`tearDown()`方法。这两个方法会分别在每调用一个测试方法的前后分别被执行  
设想你的测试需要启动一个数据库，这时，就可以在`setUp()`方法中连接数据库，在`tearDown()`方法中关闭数据库，这样，不必在每个测试方法中重复相同的代码：

```python
class TestDict(unittest.TestCase):

    def setUp(self):
        print('setUp...')

    def tearDown(self):
        print('tearDown...')
```

## 四、运行单元测试

[参考网页](https://blog.csdn.net/henni_719/article/details/56835833)
[参考网页](https://www.cnblogs.com/Mollylin/p/9765430.html)

### 方法一、
在单元测试模块的最后加上两行代码：
```python
if __name__ == '__main__':
    unittest.main()         # 全局的main（）方法。可以将一个单元测试模块变成可以直接运行的测试脚本;main()方法使用testLoader类来搜索所有包含在该模块总以“test”命名开头的测试方法，并自动执行他们
```


### （优先）方法二、
通过命令行运行：模块、类、或者单个测试方法:

> 运行测试模块：python -m unittest test_module1 test_module2   
> 运行测试类：python -m unittest test_module.TestClass   
> 运行测试方法：python -m unittest test_module.TestClass.test_method   
> ......


效果如下：
```python
PS D:\01.Software\GitHub\GitHub Repository\hello-world> python -m unittest mydict_test
E
======================================================================
ERROR: mydict_test (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: mydict_test
Traceback (most recent call last):
  File "D:\01.Software\Anaconda3\lib\unittest\loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
ModuleNotFoundError: No module named 'mydict_test'


----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
```